src/full/Agda/Interaction/BasicOps.hs

```diff
@@ -577,6 +581,47 @@ getConstraints' g f = liftTCM $ do
         let m = QuestionMark emptyMetaInfo{ metaNumber = Just $ fromIntegral ii } ii
         abstractToConcrete_ $ OutputForm noRange [] $ Assign m e
 
+-- | Goals and Warnings
+
+getGoals :: TCM Goals
+getGoals = do
+  -- visible metas (as-is)
+  visibleMetas <- typesOfVisibleMetas AsIs
+  -- hidden metas (unsolved implicit arguments simplified)
+  unsolvedNotOK <- not . optAllowUnsolved <$> pragmaOptions
+  hiddenMetas <- (guard unsolvedNotOK >>) <$> typesOfHiddenMetas Simplified
+  return (visibleMetas, hiddenMetas)
+
+getWarningsAndNonFatalErrors :: TCM WarningsAndNonFatalErrors
+getWarningsAndNonFatalErrors = do
+  mws <- getMaybeWarnings AllWarnings
+  return $ case filter isNotMeta <$> mws of
+    SomeWarnings ws@(_:_) -> swap $ classifyWarnings ws
```

Comment:
The fact this pair has type `([a],[a])` and that we sometimes mean to return it
one way and sometimes the other suggests to me that we should introduce a record
type for it.

We could move `WarningsAndNonFatalErrors` to `Agda.TypeChecking.Warnings` and
change its definition from a mere type alias to:

```haskell
data WarningsAndNonFatalErrors = WarningsAndNonFatalErrors
  { warnings       :: [TCWarning]
  , nonFatalErrors :: [TCWarning]
  }
```

and not export the constructor to force us to always use the projections...

Reply:
I'm still exporting the constructor or `WarningsAndNonFatalErrors`, because `getWarningsAndNonFatalErrors` in `Agda.TypeChecking.BasicOps` needs it.

Or should I:
1. move `getWarningsAndNonFatalErrors` to `Agda.TypeChecking.Warnings`?
2. define and export `WarningsAndNonFatalErrors [] []` as `emptyWarningsAndNonFatalErrors`?

```haskell
getWarningsAndNonFatalErrors :: TCM WarningsAndNonFatalErrors
getWarningsAndNonFatalErrors = do
  mws <- getMaybeWarnings AllWarnings
  let notMetaWarnings = filter (not . isMetaWarning) <$> mws
  return $ case notMetaWarnings of
    SomeWarnings ws@(_:_) -> classifyWarnings ws
    _ -> WarningsAndNonFatalErrors [] []
```

Reply:
2. is probably a lot simpler (`getMaybeWarnings` is defined in
`Interaction.Imports` so `getWarningsAndNonFatalErrors` would create
yet another dependency cycle)

By the way, `isMetaWarning` looks more like `isMetaTCWarning` to me.
I think it's worth defining the `isMetaWarning` and then filter using
`not . isMetaWarning . tcWarning`.